junit-platform-launcher/src/main/java/org/junit/platform/launcher/TagFilter.java

```diff
@@ -123,18 +138,36 @@ public static PostDiscoveryFilter excludeTags(String... tagExpressions) throws P
 	 * @see TestTag#isValid(String)
 	 */
 	public static PostDiscoveryFilter excludeTags(List<String> tagExpressions) throws PreconditionViolationException {
-		return includeMatching(tagExpressions, Stream::noneMatch);
+		Supplier<String> inclusionReason = () -> inclusionReasonExpressionNotSatisfy(tagExpressions);
+		Supplier<String> exclusionReason = () -> exclusionReasonExpressionSatisfy(tagExpressions);
+
+		return includeMatching(tagExpressions, Stream::noneMatch, inclusionReason, exclusionReason);
+	}
+
+	private static String inclusionReasonExpressionNotSatisfy(List<String> tagExpressions) {
+		return String.format("Test included Because it does not satisfy expression(s): [%s]",
+			formatToString(tagExpressions));
+	}
+
+	private static String exclusionReasonExpressionSatisfy(List<String> tagExpressions) {
+		return String.format("Test excluded Because it satisfy expression(s): [%s]", formatToString(tagExpressions));
+	}
+
+	private static String formatToString(List<String> tagExpressions) {
+		return tagExpressions.stream().map(String::trim).collect(Collectors.joining(","));
 	}
 
 	private static PostDiscoveryFilter includeMatching(List<String> tagExpressions,
```

Comment:
There is already a public caller method for this, common code is already abstracted out. since postDiscoveryFilter only care about the reason for inclusion/exclusion, isn't it good to just pass reasons? that's why I didn't create a new method, to keep matching logic at one place. :)

Reply:
Here the three parameters you're passing (matcher, inclusion and exclusion reasons) will always be the same trio of values, depending on wether the tags are excluded or included, so you may as well make it explicit. Plus, it will be more readable imo. Personally, I think that sometimes readability wins over abstraction ;P